graph/layout/eades.go

```diff
@@ -0,0 +1,155 @@
+// Copyright ©2019 The Gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package layout
+
+import (
+	"math"
+
+	"golang.org/x/exp/rand"
+
+	"gonum.org/v1/gonum/graph"
+	"gonum.org/v1/gonum/spatial/barneshut"
+	"gonum.org/v1/gonum/spatial/r2"
+)
+
+// EadesR2 implements the graph layout algorithm described in "A
+// heuristic for graph drawing", Congressus numerantium 42:149-160.
+// The implementation here uses the Barnes-Hut approximation for
+// global repulsion calculation and edge weights are considered
+// when calculating adjacent node attraction.
+type EadesR2 struct {
+	// M is the number of updates to perform.
+	M int
+
+	// C1, C2, C3 and C4 are the constants
+	// described in the paper.
+	C1, C2, C3, C4 float64
+
+	// Theta is the Barnes-Hut theta constant.
+	Theta float64
+
+	// Src is the source of randomness used
+	// to initialize the nodes' locations. If
+	// Src is nil, the global random number
+	// generator is used.
+	Src rand.Source
+
+	nodes   graph.Nodes
+	indexOf map[int64]int
+
+	particles []barneshut.Particle2
+	forces    []r2.Vec
+}
+
+// Update is the EadesR2 spatial graph update function.
+func (u *EadesR2) Update(g graph.Graph, layout LayoutR2) bool {
+	if u.M <= 0 {
+		return false
+	}
+	u.M--
+
+	if !layout.IsInitialized() {
+		var rnd func() float64
+		if u.Src == nil {
+			rnd = rand.Float64
+		} else {
+			rnd = rand.New(u.Src).Float64
+		}
+		u.nodes = g.Nodes()
+		u.indexOf = make(map[int64]int, u.nodes.Len())
+		u.particles = make([]barneshut.Particle2, 0, u.nodes.Len())
+		u.forces = make([]r2.Vec, u.nodes.Len())
+		for u.nodes.Next() {
+			id := u.nodes.Node().ID()
+			u.indexOf[id] = len(u.particles)
+			u.particles = append(u.particles, eadesR2Node{id: id, pos: r2.Vec{X: rnd(), Y: rnd()}})
+		}
+	}
+	u.nodes.Reset()
+
+	// Apply global repulsion.
+	plane := barneshut.NewPlane(u.particles)
+	var updated bool
+	for i, p := range u.particles {
+		f := plane.ForceOn(p, u.Theta, barneshut.Gravity2).Scale(-u.C3)
+		// Prevent marginal updates that can be caused by
+		// floating point error when nodes are very far apart.
+		if math.Hypot(f.X, f.Y) > 1e-12 {
+			updated = true
+		}
+		u.forces[i] = f
+	}
+	if !updated {
```

Comment:
What is the justification for this early return? If the nodes are too far apart, there could be attractive forces computed below. And why not having this test in the attraction loop below - if the nodes have reached an equilibrium before M runs out, Update should return false.

Reply:
I have moved this below the adjacent nodes force calculations and added conditional mutation of `updated` to that loop.